### PR TITLE
Highlights iop: X-Trans 'Reconstruct in LCh' algo rewrite

### PR DESCRIPTION
**Pull request by dtorop**
This follows up on #1188 by rewriting X-Trans 'Reconstruct in LCh' to match the algorithm @LebedevRI wrote for the Bayer variant. It is also allows for an approx. 3x speed increase.

Note that the initial commit uses a 3x3 sample for reconstruction, and the following one changes this to a 2x2 sample. The 2x2 sample seems to be an improvement, as it provides for more detail, though it also results in more artifacts at edges.

See https://imgur.com/a/cZXa1. First image compares X-Trans LCh in master with the rewrite. The second image is details processed by: 1) master, 2) PR code with 2x2 sample, and 3) the intermediate commit with a 3x3 sample.

### CODE REVIEW

**Comment by LebedevRI** [comment]
@dtorop feel free to rebase on top of current git master.
I guess it is up to you to pick 3x3 vs. 2x2 sampling.
Feel free to ping me once you think this is good to go.

On an unrelated note, i am planning (well, hoping to) work on proper pre-demosaic highlight inpainting iop sometime next.

And i only have Bayer sample raws with highlight clipping.
@dtorop, maybe you have some good x-trans raws with interestingly-clipped highlights that you can give me? Then i *might* be able to also at least try to make it work not just for Bayer...

**Comment by dtorop** [comment]
Finally getting back to this. Think have fixed the edge problem. Will test more, may be able to optimize/refine. Will write a note when it is ready to go. Also rebased this on top of git master.

@LebedevRI : apologies slow to get you a sample image for highlights. I've been using the RAF at https://app.box.com/s/cicmtfqnfny9iuqmh9d1nkt1gxa8kaw2 to test, with the xmp as given. It comes from http://www.dslr-forum.de/showpost.php?p=13800864&postcount=690. I sometimes turn on the cropping to output a particularly problematic area. Would be very excited for more work on highlights. I'm also happy to work on an X-Trans model using any Bayer code you write as a model.

**Comment by dtorop** [comment]
Have gone over this more carefully and think the code is sound. Would be curious for thoughts and if it seems ready to merge, that would be great. Made one more fix (ssize_t rather than size_t). To my surprise, this PR appears to be 2x faster than the existing X-Trans Lch code in master, and should give notably better results.

Comparison of results from this PR with extant code in dt master: http://imgur.com/a/bJPA2

File: src/iop/highlights.c [hunk 1]
```diff
@@ -531,55 +531,110 @@ static void process_lch_bayer(dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *pie
...
+          for(int offset_j = -2; offset_j <= 0; offset_j++)
           {
-            const float val = in[(size_t)jj * width + ii];
-            const int c = FCxtrans(j+jj, i+ii, roi_in, xtrans);
-            mean[c] += val;
-            cnt[c]++;
-            blend[c] = fmaxf(blend[c], (fminf(post_clip, val) - near_clip) / (post_clip - near_clip));
+            for(int offset_i = -2; offset_i <= 0; offset_i++)
+            {
+              if (clipped) {
```
Comment (LebedevRI) [inline src/iop/highlights.c]: `{` on a new line

File: src/iop/highlights.c [hunk 2]
```diff
@@ -531,55 +531,110 @@ static void process_lch_bayer(dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *pie
...
+          if(R != G && G != B)
+          {
+            const float ratio = sqrtf((Co * Co + Ho * Ho) / (C * C + H * H));
+            C *= ratio;
+            H *= ratio;
+          }
+
+          switch(FCxtrans(j, i, roi_out, xtrans))
+          {
+            case 0:
```
Comment (LebedevRI) [inline src/iop/highlights.c]: If we are really going for the speed, i wonder if just computing all three values + outputting just one will be faster, less branching? (like in `process_lch_bayer()`)

File: src/iop/highlights.c [hunk 3]
```diff
@@ -531,55 +531,110 @@ static void process_lch_bayer(dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *pie
...
-    const dt_iop_roi_t *const roi_in,
-    const uint8_t (*const xtrans)[6])
+static void process_lch_xtrans(dt_iop_module_t *self, const void *const ivoid,
+                               void *const ovoid, const dt_iop_roi_t *const roi_in,
+                               const dt_iop_roi_t *const roi_out, const float clip)
 {
+  const uint8_t (*const xtrans)[6] = self->dev->image_storage.xtrans;
+
 #ifdef _OPENMP
 #pragma omp parallel for schedule(static) default(none)
```
Comment (LebedevRI) [inline src/iop/highlights.c]: Did not do any profiling, but like with `process_lch_bayer()`, `schedule(dynamic)` **might** be better.

Comment (dtorop) [inline src/iop/highlights.c]: Yes, does look better. On my machine avg. iop execution time goes from 0.31 to 0.27 secs. (CPU time is up from 0.78 to 1.02 sec., which is OK, yes?)

File: src/iop/highlights.c (hunk 2 above)
Comment (dtorop) [inline src/iop/highlights.c]: Seems to be marginally equivalent speed (or marginally faster) to compute & output just one, but it's more concise and matches `process_lch_bayer()` so seems worth changing.

File: src/iop/highlights.c (hunk 3 above)
Comment (LebedevRI) [inline src/iop/highlights.c]: Yes, i think it just means it was able to parallelize it better, so more threads were actually running => in summary it took more time.

**Comment by LebedevRI** [comment]
Tested on my limited supply of x-trans files, looks awesome, much better than 'mode = clip', and better than current git master.
If nothing pops up, i'm ready to merge this in +24h from this comment.

File: src/iop/highlights.c (hunk 3 above)
Comment (dtorop) [inline src/iop/highlights.c]: Indeed, this must be the case. As some columns require way more work (have more highlights to reconstruct) it makes sense that with dynamic scheduling there was less chance of threads on non-clipped columns stalling out.

**Comment by dtorop** [comment]
Thank you for testing. At the risk of complicating things, I added one more change: a ring buffer to watch for the likely case in which the majority of the image is not clipped. On a test image this increases iop speed from 0.27 to 0.17 sec., while not changing the output, hence this seems worth the added complexity.

**Comment by LebedevRI** [comment]
Thanks!

**Comment by dtorop** [comment]
Thank you for checking it over, and the greatly improved Lch algorithm! I have to admit not totally understanding the math of the RGB->Lch conversion (it seems like a shorthand version of a full conversion?) but it is such a vast improvement. Glad to have it in the X-Trans code now.

**Comment by hanatos** [comment]
hehe, re: rgb->LCh .. i think that name nowadays is just legacy. nothing to
do any more with real LCh or the code in ufraw where this came from, back
then.

**Comment by dtorop** [comment]
Ah, very interesting! Have never looked at ufraw code, good to trace the lineage...

**Pull request merged**